{"cells":[{"metadata":{},"id":"automotive-stretch","cell_type":"markdown","source":"# Demo notebook for \"Finding Data for your Research Organism: Plants and RNASeq Data\""},{"metadata":{},"id":"greatest-trouble","cell_type":"markdown","source":"This notebook shows the commands to replicate the demo provided during the NCBI Webinar on August 18,2021\n\n## Workflow\n\n### 1.) Fastq files are generated from an accession list using NCBI SRA Toolkit\n### 2.) A reference genome is downloaded using NCBI Datasets command line \n### 3.) The sequences in the fastq files are aligned using NCBI's Magic-BLAST tool\n### 4.) Once files are generated, they can be loaded into the NCBI Genome Data Viewer"},{"metadata":{},"id":"composed-function","cell_type":"markdown","source":"A few notes:\n\n* This demo is intended to be run on binder as all programmatic installs have been done directly to the binder server\n* If you wish to download and use this notebook on your own server, please find the instructions for the programmatic installs in the README.md file for this git repo: https://github.com/ncbi/finding-data-demo-08-21/edit/master/README.md\n* The commands in this notebook that begin with \"!\" are bash commands that can be copied and run directly in a terminal."},{"metadata":{},"id":"visible-certification","cell_type":"markdown","source":"Command Line programs already loaded onto the server, but if you are running this notebook locally, you may need to install and add these programs to your path to work with these commands as written.\n\n**SRA Toolkit** (install instructions here: https://github.com/ncbi/sra-tools/wiki/02.-Installing-SRA-Toolkit)\n\n**Datasets Command Line** (install instructions here: https://www.ncbi.nlm.nih.gov/datasets/docs/quickstarts/command-line-tools/\n\n**Blast+ Executables** (install instructions here: https://www.ncbi.nlm.nih.gov/books/NBK279690/)\n\n**MagicBlast** (install instructions here: https://ncbi.github.io/magicblast/)"},{"metadata":{},"id":"collect-clock","cell_type":"markdown","source":"## Configure SRA Toolkit"},{"metadata":{},"id":"beneficial-lindsay","cell_type":"markdown","source":"If this is the first time running the notebook, use the following cell to set up SRA toolkit"},{"metadata":{"scrolled":true,"trusted":false},"id":"266943d1","cell_type":"code","source":"GUID = !uuidgen\nGUID = str(GUID).strip('[]')\n!vdb-config --set LIBS/GUID={GUID}\n!vdb-config --report-cloud-identity yes\n!vdb-config --set /repository/user/main/public/cache-disabled=true\n#!cat $HOME/.ncbi/user-settings.mkfg","execution_count":18,"outputs":[{"name":"stdout","output_type":"stream","text":"Report Cloud Instance Identity was set to true\r\n"}]},{"metadata":{},"id":"impressive-simpson","cell_type":"markdown","source":"Test that the sra toolkit is configured and working from this location by downloading the first two reads with quality scores using an SRA Accession."},{"metadata":{"trusted":false},"id":"continental-williams","cell_type":"code","source":"!fastq-dump --stdout -X 2 SRR3241527","execution_count":3,"outputs":[{"name":"stdout","output_type":"stream","text":"Read 2 spots for SRR3241527\r\nWritten 2 spots for SRR3241527\r\n@SRR3241527.1 FCC3E8RACXX:1:1101:8441:2000# length=100\r\nNCCACCCTTCGAAATCACAGTCCGTGAGTCTCCAACATTGGCAACATAAAGATGATCACCAACTAGCACTGCTGTTGAAGCAGTAGAACCATCATCACGG\r\n+SRR3241527.1 FCC3E8RACXX:1:1101:8441:2000# length=100\r\n#4BDFFFFHHHHHJJJIJJJHIIJGHIIEGHIJJJJJJJJIJJJJJJJJJIJJJJJJJJJJJJJJJHHHHHFFFFDFEEEEEDCCDDDDDDCDDDDDDDB\r\n@SRR3241527.2 FCC3E8RACXX:1:1101:9310:1999# length=100\r\nNGTGAGCATCCGCAGGGACCATCGCAATGCTTTGTTTTAATTAAACAGTCGGATTCCCCTTGTCCGTACCAGTTCTGAGTCGACCGTTCGACGCCCGGGG\r\n+SRR3241527.2 FCC3E8RACXX:1:1101:9310:1999# length=100\r\n#4=DDFFFHHHHHJIHJJJIJIJJJJJIIJJJJIIIJJJIJJJJJJJJFIIJJJJHHHHEFFFFFDDEDDDDCDDCDDDDDDDDDBDDDDDDDDDD@9<B\r\n"}]},{"metadata":{},"id":"insured-mouse","cell_type":"markdown","source":"## Load sequence data using SRA Toolkit command fastq-dump"},{"metadata":{},"id":"hairy-prospect","cell_type":"markdown","source":"For convenience, the accession list that we generated using the SRA Run Selector has been added to this notebook environment for you. You can reuse this notebook to run your own accessions (memory permitting) by just replacing this text file in the directory with your own list.\n\nThese 12 accessions belong to a Bioproject that measured RNA expression in Populus trichocarpus xylem on days 0, 5 and 7 of a drought treatment. More details can be found here: "},{"metadata":{"trusted":false},"id":"fb1d4801","cell_type":"code","source":"!cat accessionlist.txt","execution_count":17,"outputs":[{"name":"stdout","output_type":"stream","text":"SRR3241527\r\nSRR3241528\r\nSRR3241529\r\nSRR3241530\r\nSRR3241531\r\nSRR3241532\r\nSRR3241533\r\nSRR3241534\r\nSRR3241535\r\nSRR3241536\r\nSRR3241537\r\nSRR3241538\r\n"}]},{"metadata":{"trusted":false},"id":"5fe0ef58","cell_type":"code","source":"!cat accessionlist.txt | xargs -P 12 -n 1 fastq-dump -X 1000000 -O data","execution_count":5,"outputs":[{"name":"stdout","output_type":"stream","text":"Read 1000000 spots for SRR3241531\nWritten 1000000 spots for SRR3241531\nRead 1000000 spots for SRR3241538\nWritten 1000000 spots for SRR3241538\nRead 1000000 spots for SRR3241533\nWritten 1000000 spots for SRR3241533\nRead 1000000 spots for SRR3241532\nWritten 1000000 spots for SRR3241532\nRead 1000000 spots for SRR3241527\nWritten 1000000 spots for SRR3241527\nRead 1000000 spots for SRR3241534\nWritten 1000000 spots for SRR3241534\nRead 1000000 spots for SRR3241530\nWritten 1000000 spots for SRR3241530\nRead 1000000 spots for SRR3241529\nWritten 1000000 spots for SRR3241529\nRead 1000000 spots for SRR3241535\nWritten 1000000 spots for SRR3241535\nRead 1000000 spots for SRR3241536\nWritten 1000000 spots for SRR3241536\nRead 1000000 spots for SRR3241537\nWritten 1000000 spots for SRR3241537\nRead 1000000 spots for SRR3241528\nWritten 1000000 spots for SRR3241528\n"}]},{"metadata":{"trusted":false},"id":"surprised-transmission","cell_type":"code","source":"!ls data","execution_count":2,"outputs":[{"output_type":"stream","text":"SRR3241527.fastq  SRR3241530.fastq  SRR3241533.fastq  SRR3241536.fastq\r\nSRR3241528.fastq  SRR3241531.fastq  SRR3241534.fastq  SRR3241537.fastq\r\nSRR3241529.fastq  SRR3241532.fastq  SRR3241535.fastq  SRR3241538.fastq\r\n","name":"stdout"}]},{"metadata":{},"id":"sexual-chinese","cell_type":"markdown","source":"## Download the RefSeq Genome and annotations from Datasets\n\nThe NCBI Datasets datasets command line tools are **datasets** and **dataformat**.\n\nUse datasets to download biological sequence data across all domains of life from NCBI.\n\nUse dataformat to convert metadata from JSON Lines format to other formats.\n\nNote: The NCBI Datasets command line tools are currently in alpha and will be updated frequently to add new features, fix bugs, and enhance usability. Command syntax is subject to frequent changes. Please check back often for updates."},{"metadata":{},"id":"elementary-bolivia","cell_type":"markdown","source":"This part of the analysis repliates what we did with the web browser, except by using the command line tool provided by NCBI."},{"metadata":{"trusted":false},"id":"d5035772","cell_type":"code","source":"!datasets download genome accession GCF_000002775.4 --chromosomes 5 --include-gtf --exclude-protein --exclude-rna\n ","execution_count":8,"outputs":[{"name":"stdout","output_type":"stream","text":"Downloading: ncbi_dataset.zip    9.76kB 15.6MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    9.76kB 15.6MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    34.8kB 1.26MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    34.8kB 1.26MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    34.8kB 1.26MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    87.7kB 1.6MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    87.7kB 1.6MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    187kB 2.29MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    187kB 2.29MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    187kB 2.29MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    381kB 3.49MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    381kB 3.49MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    497kB 3.68MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    781kB 5.68MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    781kB 5.68MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    1.27MB 7.74MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    1.27MB 7.74MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    1.27MB 7.74MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    1.32MB 6.69MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    2.55MB 12.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    3.03MB 13.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    3.03MB 13.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    3.42MB 14.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    3.48MB 14.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    3.57MB 13.8MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    3.94MB 14.6MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    4.3MB 15.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    4.37MB 15.1MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    4.5MB 14.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    4.66MB 15.1MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    4.84MB 15.1MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    5.07MB 15.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    5.25MB 15.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    5.43MB 15.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    5.6MB 15.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    5.79MB 15.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    6.01MB 15.6MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    6.23MB 15.7MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    6.33MB 15.6MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    6.53MB 15.7MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    6.66MB 15.7MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    6.96MB 15.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    7.15MB 15.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    7.3MB 15.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    7.5MB 15.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    7.69MB 16MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    7.86MB 16.1MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    8.05MB 16.1MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    8.28MB 16.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    8.46MB 16.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    8.63MB 16.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    8.79MB 16.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    8.99MB 16.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    9.18MB 16.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    9.31MB 16.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    9.51MB 16.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    9.72MB 16.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    9.95MB 16.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    10.1MB 16.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    10.3MB 16.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    10.5MB 16.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    10.7MB 16.6MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    10.9MB 16.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    11.1MB 16.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    11.2MB 16.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    11.5MB 16.6MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    11.7MB 16.7MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    11.9MB 16.8MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    12.1MB 16.8MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    12.3MB 16.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    12.5MB 16.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    12.7MB 16.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    12.9MB 16.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    13.1MB 17MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    13.2MB 16.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    13.5MB 16.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    13.6MB 16.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    13.8MB 16.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    14.1MB 17.1MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    14.2MB 16.9MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    14.5MB 17.1MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    14.7MB 17.1MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    14.9MB 17.1MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    15.1MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    15.3MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    15.4MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    15.6MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    15.8MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    16MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    16.2MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    16.4MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    16.5MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    16.7MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    16.9MB 17.1MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    17.1MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    17.3MB 17.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    17.5MB 17.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    17.7MB 17.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    17.9MB 17.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    18.1MB 17.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    18.3MB 17.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    18.4MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    18.6MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    18.8MB 17.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    19MB 17.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    19.1MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    19.3MB 17.2MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    19.5MB 17.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    19.8MB 17.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    19.9MB 17.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    20.2MB 17.3MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    20.3MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    20.6MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    20.8MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    21MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    21.1MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    21.4MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    21.5MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    21.7MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    21.9MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    22.1MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    22.2MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    22.4MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    22.6MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    22.8MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    22.9MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    23.1MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    23.4MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    23.5MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    23.7MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    23.8MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    24MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    24.2MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    24.4MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    24.6MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    24.8MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    25MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    25.1MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    25.3MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    25.5MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    25.6MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    25.8MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    26.1MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    26.3MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    26.5MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    26.6MB 17.5MB/s\n"},{"name":"stdout","output_type":"stream","text":"\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    26.9MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    27.1MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    27.3MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    27.4MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    27.6MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    27.7MB 17.4MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    28MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    28.1MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    28.3MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    28.5MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    28.8MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    29MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    29.2MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    29.4MB 17.5MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    29.5MB 17.6MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    29.8MB 17.6MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    30MB 17.6MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    30.2MB 17.6MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    30.3MB 17.6MB/s\n\u001b[1A\u001b[2KDownloading: ncbi_dataset.zip    30.6MB done\nNew version of client (12.7.0) available at https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets\n"}]},{"metadata":{"trusted":false},"id":"d32c54c0","cell_type":"code","source":"!unzip -o ncbi_dataset.zip -d PtRefSeq","execution_count":9,"outputs":[{"name":"stdout","output_type":"stream","text":"Archive:  ncbi_dataset.zip\n  inflating: PtRefSeq/README.md      \n  inflating: PtRefSeq/ncbi_dataset/data/assembly_data_report.jsonl  \n  inflating: PtRefSeq/ncbi_dataset/data/GCF_000002775.4/chr5.fna  \n  inflating: PtRefSeq/ncbi_dataset/data/GCF_000002775.4/genomic.gff  \n  inflating: PtRefSeq/ncbi_dataset/data/GCF_000002775.4/genomic.gtf  \n  inflating: PtRefSeq/ncbi_dataset/data/GCF_000002775.4/sequence_report.jsonl  \n  inflating: PtRefSeq/ncbi_dataset/data/dataset_catalog.json  \n"}]},{"metadata":{},"id":"positive-baker","cell_type":"markdown","source":"Magic-BLAST requires a BLAST database in order to align the reads."},{"metadata":{"trusted":false},"id":"b8c8a1ca","cell_type":"code","source":"!makeblastdb -in PtRefSeq/ncbi_dataset/data/GCF_000002775.4/chr5.fna -dbtype nucl -parse_seqids -out PtRefSeq/Pt.chrom5\n","execution_count":10,"outputs":[{"name":"stdout","output_type":"stream","text":"\n\nBuilding a new DB, current time: 08/16/2021 18:45:42\nNew DB name:   /home/jovyan/PtRefSeq/Pt.chrom5\nNew DB title:  PtRefSeq/ncbi_dataset/data/GCF_000002775.4/chr5.fna\nSequence type: Nucleotide\nKeep MBits: T\nMaximum file size: 1000000000B\nAdding sequences from FASTA; added 1 sequences in 0.510293 seconds.\n\n\n"}]},{"metadata":{},"id":"featured-crash","cell_type":"markdown","source":"## Align Fastq Files to Reference Genome to Produce Bams using Magic-BLAST\n\nMagic-BLAST is a tool for mapping large next-generation RNA or DNA sequencing runs against a whole genome or transcriptome. Each alignment optimizes a composite score, taking into account simultaneously the two reads of a pair, and in case of RNA-seq, locating the candidate introns and adding up the score of all exons. This is very different from other versions of BLAST, where each exon is scored as a separate hit and read-pairing is ignored."},{"metadata":{},"id":"substantial-spare","cell_type":"markdown","source":"Create a location to store the bams."},{"metadata":{"trusted":false},"id":"cf4dd44f","cell_type":"code","source":"!mkdir bams","execution_count":11,"outputs":[]},{"metadata":{},"id":"revised-accent","cell_type":"markdown","source":"This next command limits the number of bases that are input into the magicblast command before running it. The default is 50M. The only reason the value is being set lower is to allow this notebook to work on the binder server, which as a 2 GB virtual memory limit. If the limit is still too high when running the notebook, try reducing the BATCH_SIZE to 5000000. "},{"metadata":{"trusted":false},"id":"aware-resistance","cell_type":"code","source":"%env BATCH_SIZE=7500000","execution_count":12,"outputs":[{"name":"stdout","output_type":"stream","text":"env: BATCH_SIZE=7500000\n"}]},{"metadata":{},"id":"differential-kruger","cell_type":"markdown","source":"Magic-BLAST has a variety of parameters that can be set. In this example, we are setting the parameters:\n\n* *splice* is *True* (transcriptomic data)\n* *no_unaligned* (to reduce the file size that is generated)\n* *max_db_word_count* is 10 to reduce virtual memory demand (Magicblast rejects 16-base word matches between input reads and the target database if this word appears in the database more than <number> of times. The default is 30.\n\nSamtools\n\nThe default output type for MagicBlast is a sam file, so we pipe it into samtools to make a bam for later viewing on the Genome Data Viewer. The sort command has a large virtual memory command.\n\n* *-m* parameter is set to 250M to reduce load in this binder notebook (it can be removed on your own machine)"},{"metadata":{"trusted":false},"id":"3c9ce308","cell_type":"code","source":"!magicblast -query data/SRR3241538.fastq -db PtRefSeq/Pt.chrom5 -splice T -no_unaligned -infmt fastq -max_db_word_count 10 | \\\nsamtools view -bS | samtools sort -m 250M -o bams/SRR3241538.bam -O BAM","execution_count":13,"outputs":[]},{"metadata":{},"id":"weekly-scheduling","cell_type":"markdown","source":"The bam files need to be indexed to be used in Genome Data Viewer."},{"metadata":{"trusted":false},"id":"3700aacd","cell_type":"code","source":"!for i in bams/*.bam; do samtools index $i; done","execution_count":15,"outputs":[]},{"metadata":{"trusted":false},"id":"minute-image","cell_type":"code","source":"!ls bams","execution_count":1,"outputs":[{"output_type":"stream","text":"SRR3241538.bam\tSRR3241538.bam.bai\r\n","name":"stdout"}]},{"metadata":{"trusted":false},"id":"85feb256","cell_type":"code","source":"#If you are using your own accession list, you can generate the list of command by running this bash command after removing the \"#\" from the beginning of each line\n\n#!cat accessionlist.txt| while read line;do \\\n#echo \"magicblast -query data/$line.fastq -db Pt.chrom5 -splice T -no_unaligned -infmt fastq -max_db_word_count 10 | \\\n#samtools view -bS | samtools sort -m 250M -o bams/$line.bam -O BAM\"; done","execution_count":14,"outputs":[]},{"metadata":{"trusted":false},"id":"orange-butler","cell_type":"code","source":"#If you wish to continue with analyzing the bam files in this notebook, here is a shortcut code. \n#Just remove the # in front of the following two lines to run them.\n\n#!cat accessionlist.txt | xargs -I % wget -O bams/%.bam https://ftp.ncbi.nlm.nih.gov/pub/education/public_webinars/2021/08Aug18_Res_Org_2/A_bams/%.bam\n#!cat accessionlist.txt | xargs -I % wget -O bams/%.bam.bai https://ftp.ncbi.nlm.nih.gov/pub/education/public_webinars/2021/08Aug18_Res_Org_2/A_bams/%.bam.bai","execution_count":null,"outputs":[]},{"metadata":{},"id":"balanced-dover","cell_type":"markdown","source":"A few of the tracks (one for Day 0, One for Day 5 and one for Day 7) have been entered and can be viewed by following this link which will be available until November 16, 2021 (90 days after creation).\n\nhttps://www.ncbi.nlm.nih.gov/genome/gdv/browser/genome/?cfg=NCID_1_8592795_130.14.18.128_9146_1629217835_4040785416\n\n\nIf the link has expired, \n\n1.) Navigate to the GDV Viewer for Populus trichocarpus chromosome 5\n\n![Navigate to Chromosome 5](navigate_GDV_Viewer.png)\n\n2.) Upon opening GDV Viewer, the bam files and their corresponding indexes can be loaded using the remote files that we have put into our ftp repository (GDV does not currently support local uploads of bam files).\n\n\nhttps://ftp.ncbi.nlm.nih.gov/pub/education/public_webinars/2021/08Aug18_Res_Org_2/A_bams/SRR3241527.bam\nhttps://ftp.ncbi.nlm.nih.gov/pub/education/public_webinars/2021/08Aug18_Res_Org_2/A_bams/SRR3241528.bam\nhttps://ftp.ncbi.nlm.nih.gov/pub/education/public_webinars/2021/08Aug18_Res_Org_2/A_bams/SRR3241529.bam\nhttps://ftp.ncbi.nlm.nih.gov/pub/education/public_webinars/2021/08Aug18_Res_Org_2/A_bams/SRR3241530.bam\nhttps://ftp.ncbi.nlm.nih.gov/pub/education/public_webinars/2021/08Aug18_Res_Org_2/A_bams/SRR3241531.bam\nhttps://ftp.ncbi.nlm.nih.gov/pub/education/public_webinars/2021/08Aug18_Res_Org_2/A_bams/SRR3241532.bam\nhttps://ftp.ncbi.nlm.nih.gov/pub/education/public_webinars/2021/08Aug18_Res_Org_2/A_bams/SRR3241533.bam\nhttps://ftp.ncbi.nlm.nih.gov/pub/education/public_webinars/2021/08Aug18_Res_Org_2/A_bams/SRR3241534.bam\nhttps://ftp.ncbi.nlm.nih.gov/pub/education/public_webinars/2021/08Aug18_Res_Org_2/A_bams/SRR3241535.bam\nhttps://ftp.ncbi.nlm.nih.gov/pub/education/public_webinars/2021/08Aug18_Res_Org_2/A_bams/SRR3241536.bam\nhttps://ftp.ncbi.nlm.nih.gov/pub/education/public_webinars/2021/08Aug18_Res_Org_2/A_bams/SRR3241537.bam\nhttps://ftp.ncbi.nlm.nih.gov/pub/education/public_webinars/2021/08Aug18_Res_Org_2/A_bams/SRR3241538.bam\n\n\n\n![How to load remote files](add_remote_track.png)\n\n\n"},{"metadata":{},"id":"exclusive-audience","cell_type":"markdown","source":"If you want to load your own remote files into GDV Viewer, it is possible using your Github repository for files that are less than 1 GB. Just use the following file format based on the location of your file in the repository.\n\nThis file has been added to this repository for testing this option.\n\nhttps://github.com/ncbi/finding-data-demo-08-21/raw/master/SRR3241537.bam\n\nMake sure the `bam` files and the `bam.bai` fils are in the same location in your repository.\n\n"},{"metadata":{"trusted":false},"id":"consistent-quilt","cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"name":"python3","display_name":"Python 3","language":"python"},"language_info":{"name":"python","version":"3.7.10","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat":4,"nbformat_minor":5}
